src/compiler/checker.ts

```diff
@@ -338,7 +338,15 @@ module ts {
                 // Locals of a source file are not in scope (because they get merged into the global symbol table)
                 if (location.locals && !isGlobalSourceFile(location)) {
                     if (result = getSymbol(location.locals, name, meaning)) {
-                        break loop;
+                        // Type parameters of a function are in scope in the entire function declaration, including the parameter
+                        // list and return type. However, local types are only in scope in the function body.
+                        if (!(meaning & SymbolFlags.Type) ||
+                            !(result.flags & (SymbolFlags.Type & ~SymbolFlags.TypeParameter)) ||
+                            !isFunctionLike(location) ||
+                            lastLocation === (<FunctionLikeDeclaration>location).body) {

```

Comment:
Initializers are evaluated *inside* the function in the generated code. We therefore need identifiers in expressions to bind to locals, even if that is subsequently deemed an error. If the locals weren't in scope at all, the following would produce no errors, but it wouldn't actually work:
```typescript
var a = 1;
function f(x = a) {  // We don't want this to bind to the global a
    let a = 2;
}
```
We don't have any such problem with local types, so we can just take them out of scope.